.ycm_extra_conf.py

```diff
@@ -1,67 +1,179 @@
+# software under copyright law.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+# IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+#
+# For more information, please refer to <http://unlicense.org/>
+
+"""
+Based on the template file provided by the 'YCM-Generator' project authored by
+Reuben D'Netto.
+Jiahui Xie has re-reformatted and expanded the original script in accordance
+to the requirements of the PEP 8 style guide and 'systemd' project,
+respectively.
+
+The original license is preserved as it is.
+"""
 
-
-def Flatten(lists):
-  return list(itertools.chain.from_iterable(lists))
+import os
+import ycm_core
 
 
 def DirectoryOfThisScript():
-  return os.path.dirname(os.path.abspath(__file__))
+    return os.path.dirname(os.path.abspath(__file__))
+
+
+def TraverseByDepth(root, include_extensions):
+    is_root = True
+    result = set()
+    # Perform a depth first top down traverse of the given directory tree.
+    for root_dir, subdirs, file_list in os.walk(root):
+        if not is_root:
+            # print("Relative Root: ", root_dir)
+            # print(subdirs)
+            if include_extensions:
+                get_ext = os.path.splitext
+                subdir_extensions = {
+                    get_ext(f)[-1] for f in file_list if get_ext(f)[-1]
+                }
+                if subdir_extensions & include_extensions:
+                    result.add(root_dir)
+            else:
+                result.add(root_dir)
+        else:
+            is_root = False
+
+    return result
+
+
+_project_src_dir = os.path.join(DirectoryOfThisScript(), "src")
+_include_dirs_set = TraverseByDepth(_project_src_dir, frozenset({".h"}))
+flags = [
+    "-x",
+    "c"
+    # The following flags are partially redundant due to the existence of
+    # 'compile_commands.json'.
+    #    '-Wall',
+    #    '-Wextra',
+    #    '-Wfloat-equal',
+    #    '-Wpointer-arith',
+    #    '-Wshadow',
+    #    '-std=gnu99',
+]
+
+for include_dir in _include_dirs_set:
+    flags.append("-I" + include_dir)
+
+# Set this to the absolute path to the folder (NOT the file!) containing the
+# compile_commands.json file to use that instead of 'flags'. See here for
+# more details: http://clang.llvm.org/docs/JSONCompilationDatabase.html
+#
+# You can get CMake to generate this file for you by adding:
+#   set( CMAKE_EXPORT_COMPILE_COMMANDS 1 )
+# to your CMakeLists.txt file.
+#
+# Most projects will NOT need to set this to anything; you can just change the
+# 'flags' list of compilation flags. Notice that YCM itself uses that approach.
+compilation_database_folder = "build/compile_commands.json"
```

Comment:
This hardcodes 'build'. Maybe that's OK, since it'll work for 80% of cases. An alternative would be a trick like https://github.com/systemd/systemd/blob/master/test/Makefile.guess does.